nexus/src/app/background/tasks/region_snapshot_replacement_step.rs

```diff
@@ -438,18 +438,125 @@ impl RegionSnapshotReplacementFindAffected {
+                }
+            };
+
+            let target_still_referenced =
+                if let Some(target_address) = &maybe_target_address {
+                    match self
+                        .datastore
+                        .volume_references_read_only_target(
+                            request.volume_id(),
+                            *target_address,
+                        )
+                        .await
+                    {
+                        Ok(maybe_referenced) => maybe_referenced.unwrap_or({
+                            // This means that the volume was deleted!
+                            false
+                        }),
+
+                        Err(e) => {
+                            let s = format!(
+                                "error determining if volume reference exists:\
+                                {e}"
+                            );
+                            error!(
+                                log,
+                                "{s}";
+                                "request_id" => %request.request_id,
+                            );
+                            status.errors.push(s);
+
+                            continue;
+                        }
+                    }
+                } else {
+                    // Note: if in this branch, then the below
+                    // `step_invalidated` check will already trip for
+                    // maybe_target_address.is_none().
+
+                    false
+                };
+
+            let step_invalidated = volume_deleted
+                || maybe_target_address.is_none()
+                || !target_still_referenced;
+
+            if step_invalidated {
+                // The replacement step was somehow invalidated, so proceed with
+                // clean up, which if this is in state Requested there won't be
+                // any additional associated state, so transition the record to
                 // Completed.
 
                 info!(
                     &log,
-                    "request {} step {} volume {} was soft or hard deleted!",
+                    "request {} step {} {}",
                     request.request_id,
                     request_step_id,
-                    request.volume_id,
+                    if volume_deleted {
+                        format!(
+                            "volume {} was soft or hard deleted!",
+                            request.volume_id,
+                        )
+                    } else if maybe_target_address.is_none() {
+                        String::from("associated read-only target was deleted")
+                    } else if !target_still_referenced {
+                        format!(
+                            "volume {} no longer references target",
+                            request.volume_id,
+                        )
+                    } else {
+                        String::from("UNKNOWN")
```

Comment:
This would happen if someone changed how `step_invalidated` was assigned and did not add a place for it here?

Reply:
Yeah, exactly